File: letter_combinations_of_a_phone_number.py
class Solution:
    def letterCombinations(self,digits):
        dic={'0':"",'1':"",'2':"abc",'3':"def",'4':"ghi",'5':"jkl",'6':"mno",'7':"pqrs",'8':"tuv",'9':"wxyz"}
        length=len(digits)
        for index in range(length):
            led
        

        
    def letterCombinations_1(self,digits):
        if not digits.isdigit():
            return []
        self.res=[]
        n=len(digits)
        self.dic={'0':"",'1':"",'2':"abc",'3':"def",'4':"ghi",'5':"jkl",'6':"mno",'7':"pqrs",'8':"tuv",'9':"wxyz"}        
        self.recursiveSearch(digits,'',0,n)
        return self.res
    def recursiveSearch(self,digits,answer,index,n):
        if index==n:
            self.res.append(answer)
            return
        for element in self.dic[digits[index]]:
            self.recursiveSearch(digits,answer+element,index+1,n)
            
if __name__=='__main__':
    sol=Solution()
    print sol.letterCombinations_1('23')

File: Algorithm/CPP/letter_combinations_of_a_phone_number.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
using namespace std;

class Solution {
	public:
		vector<string> letterCombinations(string digits)
		{

			vector<string> res;
			if (digits.size() == 0) return res;
			vector<string> dict = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
			string ans;
			dfs(res, dict, 0, digits, ans);
			return res;
		}
		void dfs(vector<string>& res, vector<string>& dict, int depth, string digits, string ans)
		{
			if (depth == digits.size())
			{
				res.push_back(ans);
				return;
			}
			for (auto chr : dict[digits[depth] - '0'])
			{
				dfs(res, dict, depth + 1, digits, ans + chr);
			}
		}
};
int  main()
{
	Solution sol;
	for (auto s : sol.letterCombinations("23")) cout << s << " ";
	cout<< endl;
	cin.get();
	return 0;
}
